py/desispec/preproc.py

```diff
@@ -0,0 +1,196 @@
+      * FITS upperlimit-inclusive vs. python upperlimit-exclusive
+      * FITS[x,y] vs. python[y,x]
+    
+    i.e. BIASSECB='[7:56,51:4146]' -> (slice(50,4146), slice(6,56))
+    '''
+    m = re.search('\[(\d+):(\d+)\,(\d+):(\d+)\]', value)
+    if m is None:
+        raise ValueError, 'unable to parse {} as [a:b, c:d]'.format(value)
+    else:
+        xmin, xmax, ymin, ymax = map(int, m.groups())
+        
+    return np.s_[ymin-1:ymax, xmin-1:xmax]
+
+def _overscan(pix, sigma=5, niter=3):
+    '''
+    returns overscan, readnoise from overscan image pixels
+    
+    Optional:
+        sigma : sigma clipping parameter for mean calculation
+        niter : number of iterative refits
+    '''
+    #- normalized median absolute deviation as robust version of RMS
+    #- see https://en.wikipedia.org/wiki/Median_absolute_deviation
+    overscan = np.median(pix)
+    absdiff = np.abs(pix - overscan)
+    readnoise = 1.4826*np.median(absdiff)
+    
+    #- input pixels are integers, so iteratively refit
+    for i in range(niter):
+        absdiff = np.abs(pix - overscan)
+        good = absdiff < sigma*readnoise
+        overscan = np.mean(pix[good])
+        readnoise = np.std(pix[good])
+    
+    return overscan, readnoise
+
+def preproc(rawimage, header, bias=False, pixflat=False, mask=False):
+    '''
+    preprocess image using metadata in header
+
+    image = ((rawimage-bias-overscan)*gain)/pixflat
+    
+    Args:
+        rawimage : 2D numpy array directly from raw data file
+        header : dict-like metadata, e.g. from FITS header, with keywords
+            CAMERA, DATE-OBS, CCDSEC, BIASSECx, DATASECx, CCDSECx
+            where x = A, B, C, D for each of the 4 amplifiers
+        
+    Optional bias, pixflat, and mask can each be:
+        False: don't apply that step
+        True: use default calibration data for that night
+        ndarray: use that array
+        filename (str or unicode): read HDU 0 and use that
+
+    Returns Image object with member variables:
+        image : 2D preprocessed image in units of electrons per pixel
+        ivar : 2D inverse variance of image
+        mask : 2D mask of image (0=good)
+        readnoise : 2D per-pixel readnoise of image
+        meta : metadata dictionary
+            TODO: define what keywords are included
+
+    preprocessing includes the following steps:
+        - bias image subtraction
+        - overscan subtraction (from BIASSEC* keyword defined regions)
+        - readnoise estimation (from BIASSEC* keyword defined regions)
+        - gain correction (from GAIN* keywords)
+        - pixel flat correction
+        - cosmic ray masking
+        - propagation of input known bad pixel mask
+        - inverse variance estimation
+        
+    Notes:
+    
+    The bias image is subtracted before any other calculation to remove any
+    non-uniformities in the overscan regions prior to calculating overscan
+    levels and readnoise.
```

Comment:
I based this on BOSS / idlspec2d, which models the 2D shape of the bias per amplifier and subtracts that prior to calculating the overscan as a remaining constant.  If the bias shape is constant in time this allows a higher statistics estimate of the remaining overscan offset vs. the noisier per-row per-exposure estimate.  See idlspec2d [bolton_biasgen.pro](http://www.sdss3.org/svn/repo/idlspec2d/trunk/pro/spec2d/bolton_biasgen.pro) and [bolton_biassub.pro](http://www.sdss3.org/svn/repo/idlspec2d/trunk/pro/spec2d/bolton_biassub.pro) (NB: these are sdss3.org links which someday may go away; I haven't found the equivalent public sdss.org link where idlspec2d now lives).  Example bias images are at http://www.sdss3.org/svn/repo/specflat/trunk/biases/ .  This method of subtracting the bias was implemented by Adam Bolton so we can ping him for details about the motivation behind this; David Schlegel @djschlegel probably knows too.

We did occasionally have electronics problems that caused time-variable biases.  These were handled as special cases for amplfiers and date ranges in the preprocessing code, falling back to row-by-row overscan subtraction as you describe.  For BOSS this was usually discovered by the quicklook equivalent as apparently large readnoise (taking the RMS of a non-flat 2D distribution).  If we use this method for DESI, we should add specific tests for non-flatness of the bias-subtracted overscan region.  For BOSS we chose *not* to automatically fall back to row-by-row overscan subtraction, since this was a symptom of actually needing humans to do something with the electronics and we wanted it to mess things up and make us notice.